project/src/main/webapp/components/document/versioning-component.js

```diff
@@ -0,0 +1,278 @@
+import {html, LitElement} from 'https://unpkg.com/@polymer/lit-element/lit-element.js?module';
+import {convertMillisToTimestamp, revisionToId, revisionFromId} from '../utils.js';
+
+export class VersioningComponent extends LitElement {
+  static get properties() {
+    return {
+      firepad: {type: Object},
+      codeMirror: {type: Object},
+      groupedRevisions: {type: Array},
+      revisionsMap: {type: Object},
+      latestSelectedHash: {type: String},
+      validCommitName: {type: Boolean},
+      commits: {type: Object},
+      show: {type: String},
+    };
+  }
+
+  constructor() {
+    super();
+    this.groupedRevisions = [];
+    this.revisionsMap = {};
+    this.latestSelectedHash = '';
+    this.validCommitName = false;
+    this.commits = [];
+    this.show = 'revisions';
+  }
+
+  // Remove shadow DOM so styles are inherited
+  createRenderRoot() {
+    return this;
+  }
+
+  closeVersioning(hasReverted) {
+    codeMirror.options.readOnly = false;
+    document.getElementById('versioning').style.display = 'none';
+    // Do not trigger init() in document.jsp otherwise firepad will save a blank pad
+    if (!hasReverted) {
+      this.dispatchEvent(new CustomEvent('close'));
+    }
+  }
+
+  getFirebaseAdapter() {
+    let firebaseAdapter = null;
+    while (firebaseAdapter == null) {
+      try {
+        firebaseAdapter = this.firepad.firebaseAdapter_;
+      } catch(e) {
+        console.log(e);
```

Comment:
If you dont need this, remove the log

Reply:
I do this because for a moment or two `this.firepad` may not be defined for this component, but it is guaranteed to eventually exist. This is to let the dev know that this value is not defined for now but to still continue trying.